connectors/sources/box.py

```diff
@@ -0,0 +1,445 @@
+            f"Connector will attempt to retry after {retry_after} seconds."
+        )
+        await self._sleeps.sleep(retry_after)
+        raise Exception("Rate limit exceeded.")
+
+    async def _handle_client_errors(self, exception):
+        match exception.status:
+            case 401:
+                await self.token._set_access_token()
+                raise
+            case 429:
+                await self._put_to_sleep(exception=exception)
+            case 404:
+                raise NotFound(f"Resource Not Found. Error: {exception}")
+            case _:
+                raise
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=NotFound,
+    )
+    async def get(self, url, headers, params=None):
+        try:
+            access_token = await self.token.get()
+            headers.update({"Authorization": f"Bearer {access_token}"})
+            return await self._http_session.get(url=url, headers=headers, params=params)
+        except ClientResponseError as exception:
+            await self._handle_client_errors(exception=exception)
+        except Exception:
+            raise
+
+    async def paginated_call(self, url, params, headers):
+        try:
+            offset = 0
+            while True:
+                params.update({"offset": offset, "limit": FETCH_LIMIT})
+                response = await self.get(url=url, headers=headers, params=params)
+                json_response = await response.json()
+                total_count = json_response.get("total_count")
+                for doc in json_response.get("entries"):
+                    yield doc
+                if offset >= total_count:
+                    break
+                offset += FETCH_LIMIT
+        except Exception:
+            raise
+
+    async def ping(self):
+        await self.get(url=ENDPOINTS["PING"], headers={})
+
+    async def close(self):
+        self._sleeps.cancel()
+        await self._http_session.close()
+
+
+class BoxDataSource(BaseDataSource):
+    name = "Box"
+    service_type = "box"
+
+    def __init__(self, configuration):
+        super().__init__(configuration=configuration)
+        self.configuration = configuration
+        self.tasks = 0
+        self.is_enterprise = configuration["is_enterprise"]
+        self.queue = MemQueue(maxmemsize=QUEUE_MEM_SIZE, refresh_timeout=120)
+        self.fetchers = ConcurrentTasks(max_concurrency=MAX_CONCURRENCY)
+
+    def _set_internal_logger(self):
+        self.client.set_logger(logger_=self._logger)
+
+    @cached_property
+    def client(self):
+        return BoxClient(configuration=self.configuration)
+
+    def tweak_bulk_options(self, options):
+        """Tweak bulk options as per concurrent downloads support by Box
+
+        Args:
+            options (dict): Config bulker options.
+        """
+        options["concurrent_downloads"] = CONCURRENT_DOWNLOADS
```

Comment:
What is the reason for rewriting the bulk options?

Reply:
Box has 1000 API requests per minute. So at a time, we can download approximately 15 files. So, I have changed the bulk options to 15.

Reply:
Concurrent downloads can be part of the global framework configuration or connector configuration. It cannot be a "hardcoded" constant. If you want a limit, you should introduce `MAX_CONCURRENT_DOWNLOADS` and lower the number if it's higher than the limit.